components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
+                                         DOMString::from_string("".to_string());
+                let den_img_src_not_found = source_set.image_sources.iter().filter(|imgsource|
+                    match imgsource.descriptor.den {
+                        Some(x) => (x == (1 as f64)),
+                        _ => false
+                    }
+                ).count() == 0;
+                let wid_img_src_not_found = source_set.image_sources.iter().filter(|imgsource|
+                    match imgsource.descriptor.wid {
+                        Some(_x) => true,
+                        _ => false
+                    }
+                ).count() == 0;
+                if !is_str_empty && den_img_src_not_found && wid_img_src_not_found {
+                    source_set.image_sources.push(ImageSource {
+                        url: element.get_string_attribute(&local_name!("src")).to_string(),
+                        descriptor: Descriptor { wid: None, den: None }
+                    })
+                }
+
+                // Step 4.1.4
+                Self::normalise_source_densities(&mut source_set);
+
+                // Step 4.1.5
+                *self.source_set.borrow_mut() = source_set;
+
+                // Step 4.1.6
+                return ;
+            }
+            // Step 4.2
+            if !element.is::<HTMLSourceElement>() {
+                continue;
+            }
+
+            // Step 4.3 - 4.4
+            let mut source_set = HTMLImageElement::new_source_set();
+            match element.get_attribute(&ns!(), &local_name!("srcset")) {
+                Some(x) => {
+                    source_set.image_sources = parse_a_srcset_attribute(&x.value());
+                }
+                _ => continue
+            }
+
+            // Step 4.5
+            if source_set.image_sources.len() == 0 {
+                continue;
+            }
+
+            // Step 4.6
+            match element.get_attribute(&ns!(), &local_name!("media")) {
+                Some(x) => {
+                    // Check if value of the attribute matches the environment
+                    if self.matches_environment(x.value().to_string()) {
+                        ()
+                    } else {
+                        continue;
+                    }
+                }
+                _ => ()
+            }
+
+            // Step 4.7
+            match element.get_attribute(&ns!(), &local_name!("sizes")) {
+                Some(x) => {
+                    source_set.source_size =
+                        parse_a_sizes_attribute(DOMString::from_string(x.value().to_string()));
+                    self.set_value_source_size_list(& mut source_set.source_size);
+                }
+                _ => ()
+            }
+
+            // Step 4.8
+            match element.get_attribute(&ns!(), &local_name!("type")) {
+                Some(x) => {
+                    match get_mime_type_str(&x.value()) {
```

Comment:
get_mime_type_str doesn't do what we want here, since it returns the mime type matching a provided file extension. I'm not sure of the right way to implement this right now, so just put a TODO comment about comparing the mime type against a list of supported image mime types.

Reply:
I will add a todo. Also, should I remove match get_mime_type_str ?

Reply:
Yes.